libbeat/publisher/queue/memqueue/eventloop.go

```diff
@@ -203,49 +207,25 @@ func (l *directEventLoop) processACK(lst chanList, N int) {
 	entries := l.buf.entries
 
 	firstIndex := lst.front().start
-	// Position the index at the end of the block of ACKed events
-	idx := (firstIndex + N - 1) % len(entries)
-
-	total := 0
+	// We traverse the block of events newest to oldest so we encounter the
+	// highest produer IDs first and can skip subsequent callbacks to the
+	// same producer.
 	for i := N - 1; i >= 0; i-- {
-		if idx < 0 {
-			idx = len(entries) - 1
-		}
-
-		client := &entries[idx].client
-		log.Debugf("try ack index: (idx=%v, i=%v, seq=%v)\n", idx, i, client.seq)
-
-		idx--
-		if client.state == nil {
-			log.Debug("no state set")
-			continue
-		}
-
-		count := (client.seq - client.state.lastACK)
-		if count == 0 || count > math.MaxUint32/2 {
-			// seq number comparison did underflow. This happens only if st.seq has
-			// already been acknowledged
-			// log.Debug("seq number already acked: ", st.seq)
-
-			client.state = nil
+		idx := (firstIndex + i) % len(entries)
```

Comment:
It's not obvious what's going on with indexing here, could you please add a comment explaining this?

Also, I don't see where `len(entries)` is changed in the loop, it might worth to just pre-calculate this into `i` when the loop starts.

Reply:
`len(entries)` doesn't change, only `i` does. I've added more comments clarifying where this value is coming from, see if that makes more sense.